Add training nb_epoch to hyperparameter optimization

cjgalvin opened the pull request:
# Pull Request Template

## Description
Adds nb_epoch argument to specify number of training epochs for applicable models to hyperparam_search function for Grid Search (grid_search.py) and Gaussian Process Search (gaussian_process.py). 

Fix #2194

For Gaussian Search:
1) Add `nb_epochs` as an argument to the `hyperparam_search` class function in `gaussian_process.py`
2) Defined a new class function called `_optimize`, which takes `nb_epochs` and '**placeholders` as arguments
3) Moved the contents of the `optimizing_function` function to `_optimize`, and rewrote `optimizing_function` to return `_optimize(nb_epochs, **placeholders)`
4) Changed the `model.fit(train_dataset)` call to a `try except` statement that first tries `model.fit(train_dataset, nb_epoch=nb_epoch)`, and on a `TypeError` exception calls `model.fit(train_dataset)`. This change handles model types that do not accept an nb_epoch argument.

For Grid Search:
1) Add `nb_epochs` as an argument to the `hyperparam_search` class function in `grid_process.py`
2) Changed the `model.fit(train_dataset)` call to a `try except` statement that first tries `model.fit(train_dataset, nb_epoch=nb_epoch)`, and on a `TypeError` exception calls `model.fit(train_dataset)`. This change handles model types that do not accept an nb_epoch argument.

For Base Class:
1) Added `nb_epoch` to `hyperparam_search` function in `base_classes`.py for completeness.

Grid Search does not use a private optimizing function, so there was no need to make the same change as in Gaussian Search.

Added unit tests for Gaussian Process search and Grid Search, both of which specify `nb_epoch` as an integer different from the default in the `hyperparam_search` function call. However, these tests are not a direct test of the number of training epochs ran, and instead only test that hyperparam_search function can be called with the nb_epoch argument. I could not find a reasonable way to confirm the number of training epochs. Suggestions are welcome. I did run tests by changing `nb_epoch` and looking at the training results, confirming expected changes in training logging output.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [o] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [o] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [o] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [?] Run `mypy -p deepchem` and check no errors
  - [o] Run `flake8 <modified file> --count` and check no errors
- [o] I have performed a self-review of my own code
- [o] I have commented my code, particularly in hard-to-understand areas
- [o] I have made corresponding changes to the documentation
- [?] I have added tests that prove my fix is effective or that my feature works
- [o] New unit tests pass locally with my changes
- [o] I have checked my code and corrected any misspellings

? `mypy -p deepchem` gives me an error in `grid_search.py` that complains that model.fit() does not take nb_epoch argument, citing the `Model` class. The `fit` method of `Model` does not take the nb_epoch argument. However, some classes that subclass `Model` do take an `nb_epoch` argument. The `try except` clause should handle this situation. This error does not come up in `gaussian_process.py`, so not sure why it does in `grid_search.py`. Open to suggestions on how to proceed.

? See discussion of unit tests in the Discussion

peastman commented:
This looks good!  The one change I'd suggest is to make the default number of epochs be 10, since that will match the current behavior.

I'm not sure why mypy complains about one class and not the other, since the code seems to be identical.  It certainly is correct in calling that an error, since the base class doesn't have the argument, but in practice the try/except should handle it.  If all else fails, you can add the type comment `# type: ignore` to that line.

rbharath commented on deepchem/hyper/gaussian_process.py:
```diff
@@ -277,7 +284,11 @@ def optimizing_function(**placeholders):
       model = self.model_builder(**hyper_parameters)
-      model.fit(train_dataset)
+      try:
+        model.fit(train_dataset, nb_epoch=nb_epoch)
```
One question that comes to mind is whether this snippet will work for models that don't have `nb_epoch` defined. Like for the random forest models. I think our other tests for hyperparam opt are still passing so we're probably all right but might be good to double check.

cjgalvin replied:
Locally, the tests for hyperparam opt pass. To be sure, I adapted the random forest model used in `test_gaussian_hyperparam_opt.py` and `test_grid_hyperparam_opt.py` to a stand alone script, and tested behavior with and without specifying `nb_epoch`. The optimization ran in both cases, and returned the correct `best_hyperparams` and `best_model` (i.e., the one with the highest pearson_r2_score). Setting `nb_epoch` to some absurdly high value did not affect execution time, so it is being ignored during optimization.

If there is a standard dataset to test optimization with, let me know and I'll drop that dataset in to test. I used the same randomly generated Numpy arrays, so not a good test of actual optimization.

rbharath replied:
Ah great, thank you for the extra checks! That should work for this PR.

We don't currently have a standard dataset to test optimization with right now. It's a little tricky since the smallest molnet datasets are ~1K compounds and would be a little too slow for the test suite. Perhaps we should use something like Iris or tiny dataset as a standard benchmark though. Might be a good avenue for future improvements

rbharath reviewed (commented):
Looks good! +1 to @peastman's suggestion to set default at 10 epochs and a quick question about whether the sklearn/rf/other models still hyperparameter optimize correctly. Besides those, this looks good togo

peastman commented:
Given that lots of subclasses add other arguments to `fit()`, I wonder if the base class version should include `**kwargs`?  Or would mypy then complain about subclasses not also accepting arbitrary arguments?

cjgalvin commented:
i changed default value to 10.

Locally, I added `# type: ignore` to the line causing the `mypy` error, and it passed. 
Do you want me to add that to the repo and push?

rbharath commented on deepchem/hyper/base_classes.py:
```diff
@@ -110,6 +111,8 @@ def hyperparam_search(self,
       `train_dataset` and `valid_dataset` may have been transformed
       for learning and need the transform to be inverted before
       the metric can be evaluated on a model.
+    nb_epoch: int, (default 5)
```
Typo: Could you update the default in the docstring?

rbharath commented on deepchem/hyper/gaussian_process.py:
```diff
@@ -160,6 +161,9 @@ def hyperparam_search(self,
       `train_dataset` and `valid_dataset` may have been transformed
       for learning and need the transform to be inverted before
       the metric can be evaluated on a model.
+    nb_epoch: int, (default 5)
```
Same here

rbharath reviewed (commented):
Have a minor comment about typos but otherwise looking good now!

rbharath commented:
I think for now adding the `# type: ignore` is a decent workaround. If it turns out longer term that there are "meta-hyperparameters" like `nb_epoch`, we may want to generalize this further

rbharath commented:
With latest changes, looks good on my end! We can merge in once @peastman has had a chance to take a look

peastman commented:
Looks good!

rbharath commented:
Going to merge in. Congrats on your first merged PR @cjgalvin! Hope to see many more coming our way :)

This PR will be in DeepChem 2.5.0 which we are targeting for the end of February

cjgalvin commented:
@rbharath Thank you, glad to contribute to such a useful project 😄 
Looking forward to 2.5.0
